### ISSUE TITLE
Too many /_indom requests when chart has many time-series/instances

### ISSUE THREAD

**Issue by meteorfox** [Netflix/vector#21]
When loading a widget that pulls multiple time-series (i.e. Per-CPU utilization, Network Throughput), for each time-series an HTTP request is sent with this format:

`http://<host>:44323/pmapi/<context>/_indom?name=<metric>&instance=<instance_num>`

This slows down significantly the browser, and the widgets remain with 'No available data' for long times. 

For example, one of my hosts has 12 CPUs, the CPU utilization fires 24 HTTP requests to gather the name of the time-series/instance, 12 times for `kernel.percpu.cpu.sys`, and 12 times for `kernel.percpu.cpu.user`.

Here's a screenshot of network tab in Chrome:
![screenshot from 2015-04-09 17 33 02](https://cloud.githubusercontent.com/assets/232300/7078453/cc620870-dede-11e4-9bc8-26951ff3a866.png)

For Network throughput is even worse, since pmwebd is slow retrieving network data and I have 25 network interfaces (bridges, bonds, and interfaces), I can't even load the chart, unless I wait minutes for it.

The culprit seems to be how `Metric.pushValues()` and `MetricService.getInames()` functions are designed, and how they are called by `MetricListService.updateMetrics()`, instead of getting all Inames in one batch, it retrieves them one by one.

It seems that pmwebd supports retrieving all the instances name in one batch by doing just dropping the `instance` query parameter like so:

`GET http://<host>:44323/pmapi/<context>/_indom?name=kernel.percpu.cpu.user`

with returned the following response:

```json
{
    "indom":251658240,
    "instances":[
        { "instance":0, "name":"cpu0" },
        { "instance":1, "name":"cpu1" },
        { "instance":2, "name":"cpu2" },
        { "instance":3,  "name":"cpu3" },
        { "instance":4,  "name":"cpu4" },
        { "instance":5,  "name":"cpu5" },
        { "instance":6,  "name":"cpu6" },
        { "instance":7,  "name":"cpu7" },
        { "instance":8,  "name":"cpu8" },
        { "instance":9,  "name":"cpu9" },
        { "instance":10,  "name":"cpu10" },
        { "instance":11,  "name":"cpu11" }
    ]
}
```

**Comment by spiermar** [comment]
Merged. Testing here.

**Comment by meteorfox** [comment]
That pull request was not complete, while it doesn't break anything it still hasn't addressed the perf issues with the multiple /_indom requests

**Comment by spiermar** [comment]
Create a new branch (https://github.com/Netflix/vector/tree/inames_batching). Feel free to send pull requests to it.

**Comment by meteorfox** [comment]
Awesome thanks.

**Comment by spiermar** [comment]
Merged. Will test it on Monday.

**Comment by spiermar** [comment]
Something wrong with the _indom requests. Debugging later today.

**Comment by meteorfox** [comment]
Thanks for testing, and sorry I didn't catch that bug. If you can give me more details on how to recreate it, I could help with the debugging.

I did notice that in my environment pmwebd sometimes fails with 503, and the process dies, when using the batch method. Then I have to restart pmwebd again, and it starts working. This issue, if the same, would be a pmwebd bug, not Vector.

**Comment by fche** [comment]
@meteorfox, please nail a gdb to a dying pmwebd / coredump and collect a traceback.  This should not happen.

**Comment by meteorfox** [comment]
@fche ok, I'll open a new issue to track that. If that problem ends up being the same I'm having then we can just reference it.

**Comment by meteorfox** [comment]
@fche Opened issue #31 to track it

**Comment by spiermar** [comment]
@meteorfox The problem I'm seeing is basically a HTTP 400 for the _indom requests

Remote Address:54.176.222.137:7402
Request URL:http://pcpdemo.test.netflix.net:7402/pmapi/823341773/_indom?instance&name=hinv.ncpu
Request Method:GET
Status Code:400 Bad Request

PMWEBAPI error, code -12359: Unknown or illegal instance domain identifier

Basically missing the =0 on the instance parameter.

**Comment by meteorfox** [comment]
@spiermar Can you show me the response for this request:

`GET http://localhost:44323/pmapi/<context_id>/_fetch?names=hinv.ncpu`

I built PCP from source v3.10+, and I get this response from below,
```json
{
    "timestamp":{
        "s":1428959511,
        "us":573329 
     }, 
    "values":[
         {
               "pmid":251658272,
               "name":"hinv.ncpu",
               "instances":[{"instance":-1, "value":8 }]
         }
     ]
}
```
 which then it takes the `-1` from instance and uses it as the `instance` query param for  /_indom. Like so `http://localhost:44323/pmapi/102211716/_indom?instance=-1&name=hinv.ncpu`

Perhaps, your version returns an empty string or something, and my code produces that bad request. This seems like a quick fix, I just need to see your response to make sure I handle that case.

**Comment by spiermar** [comment]
GET http://pcpdemo.test.netflix.net:7402/pmapi/1965300578/_fetch?names=hinv.ncpu

{ "timestamp": { "s":1428960706, "us":670939 },
"values": [
{ "pmid":251658272, "name":"hinv.ncpu","instances": [
{"value":2}] }] }

**Comment by spiermar** [comment]
No "instance" parameter in the "instances" array.

**Comment by meteorfox** [comment]
That pull-request should fix the problem.

**Comment by spiermar** [comment]
Looks good @meteorfox !

**Comment by meteorfox** [comment]
Thanks for taking the time to review it and taking the pull request. 

I think now that the PMAPIService is in place, we could start making other Services/Factories/Controllers make use of it, instead of them directly making the HTTP requests. I changed a few places already, but I don't think I got them all. Also, there are few places were code could be simplified because of this new service.

Unless you have other suggestions, or ideas where to move next, I could start working on a new pull-request.
